numba/targets/rangeobj.py

```diff
@@ -65,6 +65,25 @@ def range3_impl(context, builder, sig, args):
                                   range_state_type,
                                   state._getvalue())
 
+    @lower_builtin(len, range_state_type)
+    def range_len(context, builder, sig, args):
+        """
+        len(range)
+        """
+        (value,) = args
+        state = RangeState(context, builder, value)
+        res = RangeIter.from_range_state(context, builder, state)
+        return impl_ret_untracked(context, builder, int_type, builder.load(res.count))
+
+    @lower_builtin(len, range_iter_type)
+    def range_iter_len(context, builder, sig, args):
```

Comment:
@ninegua, it is possible to implement the `len_iter_range` using the `intrinsic` in the extending API.   It is not yet documented in the user docs but it has a very detail docstring; see https://github.com/numba/numba/blob/master/numba/extending.py#L286.   More example usage is available at the unit test  in https://github.com/numba/numba/blob/master/numba/tests/test_extending.py#L533.  With it, you can access the IRbuilder just like you did with `lower_builtin`.  That way, you can avoid writing an extra pass and avoid conflict the python semantic.